src/SdkCommon/ClientRuntime/ClientRuntime.Tests/PolymorphicJsonConverterTests.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using Xunit;
+using Newtonsoft.Json;
+using Microsoft.Rest.Serialization;
+
+namespace Microsoft.Rest.ClientRuntime.Tests
+{
+    public class PolymorphicJsonConverterTests
+    {
+        // Example hierarchy
+        // Naming convention:  nameof(U).StartsWith(nameof(T))  <=>  U : T
+
+        private class Model { }
+        private class ModelA : Model { }
+        private class ModelB : Model { }
+        private class ModelAA : ModelA { }
+        private class ModelAB : ModelA { }
+        private class ModelABA : ModelAB { }
+        private class ModelABB : ModelAB { }
+
+        // JsonConverters & helpers
+
+        private static JsonConverter SerializeJsonConverter => new PolymorphicSerializeJsonConverter<Model>("type");
+        private static JsonConverter DeserializeJsonConverter => new PolymorphicDeserializeJsonConverter<Model>("type");
+        private static T RoundTrip<T>(T instance) where T : Model
+        {
+            var json = JsonConvert.SerializeObject(instance, SerializeJsonConverter);
+            return JsonConvert.DeserializeObject<T>(json, DeserializeJsonConverter);
+        }
+        private static void AssertRoundTrips<T>(T instance) where T : Model
```

Comment:
@olydis do we have any negative tests, where it fails to convert if we reverse the order of inheritance hierarchy? If not we should add those.